sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/internal/SubscribeTransform.java

```diff
@@ -101,6 +91,42 @@ private MemoryBufferedSubscriber newBufferedSubscriber(
         consumer -> newSubscriber(subscriptionPartition.partition(), startOffset, consumer));
   }
 
+  private MemoryBufferedSubscriber getCachedSubscriber(
+      SubscriptionPartition subscriptionPartition, Offset startOffset) {
+    Supplier<MemoryBufferedSubscriber> getOrCreate =
+        () ->
+            PerServerSubscriberCache.CACHE.get(
+                subscriptionPartition,
+                () -> newBufferedSubscriber(subscriptionPartition, startOffset));
+    while (true) {
+      MemoryBufferedSubscriber subscriber = getOrCreate.get();
+      Offset fetchOffset = subscriber.fetchOffset();
+      if (startOffset.equals(fetchOffset)) {
+        return subscriber;
+      }
+      LOG.info(
+          "Discarding subscriber due to mismatch, this should be rare. {}, start: {} fetch: {}",
+          subscriptionPartition,
+          startOffset,
+          fetchOffset);
+      try {
+        subscriber.stopAsync().awaitTerminated();
```

Comment:
does the get above remove it from the cache? Or do you have to do so here so a new one is returned next time?

Reply:
The cache uses service shutdown to force evict from the cache. So yeah this will get a new one https://github.com/apache/beam/blob/0943d85113e77505f2a1291d7b256b9a4b1e5dec/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/internal/ServiceCache.java#L60